Mathlib/Algebra/Ring/ULift.lean

```diff
@@ -61,123 +60,50 @@ theorem down_ofNat [NatCast R] (n : ℕ) [n.AtLeastTwo] :
-  { ULift.one, ULift.addMonoid with
-      natCast := (⟨·⟩)
-      natCast_zero := congr_arg ULift.up Nat.cast_zero,
-      natCast_succ := fun _ => congr_arg ULift.up (Nat.cast_succ _) }
+instance addMonoidWithOne [AddMonoidWithOne R] : AddMonoidWithOne (ULift R) where
+  natCast_zero := congr_arg ULift.up Nat.cast_zero
+  natCast_succ _ := congr_arg ULift.up (Nat.cast_succ _)
 
-instance addCommMonoidWithOne [AddCommMonoidWithOne R] : AddCommMonoidWithOne (ULift R) :=
-  { ULift.addMonoidWithOne, ULift.addCommMonoid with }
+instance addCommMonoidWithOne [AddCommMonoidWithOne R] : AddCommMonoidWithOne (ULift R) where
 
-instance addGroupWithOne [AddGroupWithOne R] : AddGroupWithOne (ULift R) :=
-  { ULift.addMonoidWithOne, ULift.addGroup with
-      intCast := (⟨·⟩),
-      intCast_ofNat := fun _ => congr_arg ULift.up (Int.cast_natCast _),
-      intCast_negSucc := fun _ => congr_arg ULift.up (Int.cast_negSucc _) }
+instance addGroupWithOne [AddGroupWithOne R] : AddGroupWithOne (ULift R) where
+  intCast_ofNat _ := congr_arg ULift.up (Int.cast_natCast _)
+  intCast_negSucc _ := congr_arg ULift.up (Int.cast_negSucc _)
 
-instance addCommGroupWithOne [AddCommGroupWithOne R] : AddCommGroupWithOne (ULift R) :=
-  { ULift.addGroupWithOne, ULift.addCommGroup with }
+instance addCommGroupWithOne [AddCommGroupWithOne R] : AddCommGroupWithOne (ULift R) where
 
 instance nonUnitalNonAssocSemiring [NonUnitalNonAssocSemiring R] :
-    NonUnitalNonAssocSemiring (ULift R) :=
-  { zero := (0 : ULift R), add := (· + ·), mul := (· * ·), nsmul := AddMonoid.nsmul,
-    zero_add, add_zero, zero_mul, mul_zero, left_distrib, right_distrib,
-    nsmul_zero := fun _ => AddMonoid.nsmul_zero _,
-    nsmul_succ := fun _ _ => AddMonoid.nsmul_succ _ _,
-    add_assoc, add_comm }
-
-instance nonAssocSemiring [NonAssocSemiring R] : NonAssocSemiring (ULift R) :=
-  { ULift.addMonoidWithOne with
-      nsmul := AddMonoid.nsmul, natCast := fun n => ULift.up n, add_comm, left_distrib,
-      right_distrib, zero_mul, mul_zero, one_mul, mul_one }
-
-instance nonUnitalSemiring [NonUnitalSemiring R] : NonUnitalSemiring (ULift R) :=
-  { zero := (0 : ULift R), add := (· + ·), mul := (· * ·), nsmul := AddMonoid.nsmul,
-    add_assoc, zero_add, add_zero, add_comm, left_distrib, right_distrib, zero_mul, mul_zero,
-    mul_assoc, nsmul_zero := fun _ => AddMonoid.nsmul_zero _,
-    nsmul_succ := fun _ _ => AddMonoid.nsmul_succ _ _ }
-
-instance semiring [Semiring R] : Semiring (ULift R) :=
-  { ULift.addMonoidWithOne with
-      nsmul := AddMonoid.nsmul,
-      npow := Monoid.npow, natCast := fun n => ULift.up n, add_comm, left_distrib, right_distrib,
-      zero_mul, mul_zero, mul_assoc, one_mul, mul_one, npow_zero := fun _ => Monoid.npow_zero _,
-      npow_succ := fun _ _ => Monoid.npow_succ _ _ }
+    NonUnitalNonAssocSemiring (ULift R) where
+
+instance nonAssocSemiring [NonAssocSemiring R] : NonAssocSemiring (ULift R) where
```

Comment:
Why keep these at all? Is it priority? Is it valuable to record some constant capturing such a term? Writing down things Lean can already figure out is what dragged us down. I would think the default would be to not do so unless there is some solid justification.

Reply:
You don't find this syntax slightly mental for what is going on? `deriving` exists for exactly this space: automatic instances that synthesis can't find. It also has the benefit that a user can get a much better appreciation of the content from the syntax. You can also chain them in one line or so.

Reply:
Strongly disagree for many reasons but this is on tangent. I mainly wanted to record my thought that exposing the elaboration of where would make for a nice universal deriving handler.